lib/commands/network.js

```diff
@@ -37,6 +40,102 @@ commands.isWifiOn = async function isWifiOn () {
+ * @throws {Error} If none of known properties were provided or there was an error
+ * while changing connectivity states
+ */
+commands.mobileSetConnectivity = async function mobileSetConnectivity (opts = {}) {
+  const {
+    wifi,
+    data,
+    airplaneMode,
+  } = opts;
+
+  if (_.every([wifi, data, airplaneMode], _.isUndefined)) {
+    throw new errors.InvalidArgumentError(
+      `Either one of ['${WIFI_KEY_NAME}', '${DATA_KEY_NAME}', '${AIRPLANE_MODE_KEY_NAME}'] options must be provided`
+    );
+  }
+
+  const currentState = await this.mobileGetConnectivity({
+    services: [
+      ...(_.isUndefined(wifi) ? [] : [WIFI_KEY_NAME]),
+      ...(_.isUndefined(data) ? [] : [DATA_KEY_NAME]),
+      ...(_.isUndefined(airplaneMode) ? [] : [AIRPLANE_MODE_KEY_NAME]),
+    ]
+  });
+  const setters = [];
+  if (!_.isUndefined(wifi) && currentState.wifi !== Boolean(wifi)) {
+    setters.push(this.adb.setWifiState(wifi, this.isEmulator()));
+  }
+  if (!_.isUndefined(data) && currentState.data !== Boolean(data)) {
+    setters.push(this.adb.setDataState(data, this.isEmulator()));
+  }
+  if (!_.isUndefined(airplaneMode) && currentState.airplaneMode !== Boolean(airplaneMode)) {
+    setters.push(async () => {
+      await this.adb.setAirplaneMode(airplaneMode);
+      if (this.adb.getApiLevel() < 30) {
+        await this.adb.broadcastAirplaneMode(airplaneMode);
+      }
+    });
+  }
+  if (!_.isEmpty(setters)) {
+    await B.all(setters);
+  }
+};
+
+/**
+ * @typedef {Object} GetConnectivityResult
+ * @property {boolean} wifi True if wifi is enabled
+ * @property {boolean} data True if mobile data connection is enabled
+ * @property {boolean} airplaneMode True if Airplane Mode is enabled
+ */
+
+/**
+ * @typedef {Object} GetConnectivityOptions
+ * @property {string[]|string?} services one or more services to get the connectivity for.
+ * Supported service names are: wifi, data, airplaneMode.
+ */
+
+/**
+ * Retrieves the connectivity properties from the device under test
+ *
+ * @param {GetConnectivityOptions?} opts If no service names are provided then the
+ * connectivity state is returned for all of them.
+ * @returns {GetConnectivityResult}
+ */
+commands.mobileGetConnectivity = async function mobileGetConnectivity (opts = {}) {
+  let {
+    services = [WIFI_KEY_NAME, DATA_KEY_NAME, AIRPLANE_MODE_KEY_NAME],
+  } = opts;
+  if (!_.isArray(services)) {
+    services = [services];
+  }
+
+  const statePromises = {
+    wifi: B.resolve(services.includes(WIFI_KEY_NAME) ? undefined : this.adb.isWifiOn()),
+    data: B.resolve(services.includes(DATA_KEY_NAME) ? undefined : this.adb.isDataOn()),
+    airplaneMode: B.resolve(
+      services.includes(AIRPLANE_MODE_KEY_NAME) ? undefined : this.adb.isAirplaneModeOn()
+    ),
+  };
+  await B.all(_.values(statePromises));
+  return _.fromPairs(services.map((k) => [k, statePromises[k].value()]));
+};
+
 commands.setNetworkConnection = async function setNetworkConnection (type) {
```

Comment:
should we rewrite the implementation of this method to use the new ADB methods?

Reply:
maybe...
I don't like it much because it uses a non-standard endpoint and should be deprecated in favour of an extension command

Reply:
this endpoint is "standard" MJSONWP (https://github.com/SeleniumHQ/mobile-spec/blob/master/spec-draft.md#device-modes) and supported in all the clients i think. not sure if that changes your mind at all or if you intend to move every non-w3c endpoint to an extension command?

Reply:
> if you intend to move every non-w3c endpoint to an extension command?

This is exactly what I wanted

Reply:
OK carry on then! We should follow a plan where the execute method is released, then clients are updated to simply wrap the execute method, then we can remove the additional route.